File: Examples/objective-c/Showroom/Showroom/Estimote/ShowroomManager.m
//
//  ShowroomManager.m
//  Showroom
//

#import "ShowroomManager.h"

#import <EstimoteSDK/EstimoteSDK.h>

@interface ShowroomManager () <ESTTriggerManagerDelegate>

@property (nonatomic) NSDictionary<NearableID *,Product *> *products;

@property (nonatomic) ESTTriggerManager *triggerManager;
@property (nonatomic) NSMutableArray<ESTTrigger *> *triggers;

@end

@implementation ShowroomManager

- (instancetype)initWithProducts:(NSDictionary<NearableID * ,Product *> *)products {
    self = [super init];
    if (self) {
        _products = products;

        self.triggerManager = [ESTTriggerManager new];
        self.triggerManager.delegate = self;

        self.triggers = [NSMutableArray new];

        for (NearableID *nearableID in self.products.keyEnumerator) {
            ESTMotionRule *motionRule = [ESTMotionRule motionStateEquals:YES forNearableIdentifier:nearableID.nearableIDString];
            ESTTrigger *motionTrigger = [[ESTTrigger alloc] initWithRules:@[motionRule] identifier:nearableID.nearableIDString];
            [self.triggers addObject:motionTrigger];
        }
    }
    return self;
}

- (void)startUpdates {
    for (ESTTrigger *trigger in self.triggers) {
        [self.triggerManager startMonitoringForTrigger:trigger];
    }
}

- (void)stopUpdates {
    for (ESTTrigger *trigger in self.triggers) {
        [self.triggerManager stopMonitoringForTriggerWithIdentifier:trigger.identifier];
    }
}

- (void)triggerManager:(ESTTriggerManager *)manager triggerChangedState:(ESTTrigger *)trigger {
    NearableID *nearableID = [[NearableID alloc] initWithNearableIDString:trigger.identifier];
    if (!nearableID) { return; }

    Product *product = [self.products objectForKey:nearableID];
    if (!product) { return; }

    if (trigger.state == YES) {
        [self.delegate showroomManager:self didDetectPickupForProduct:product];
    } else {
        [self.delegate showroomManager:self didDetectPutdownForProduct:product];
    }
}

@end

File: Examples/swift/Showroom/Showroom/Estimote/ShowroomManager.swift
//
// Please report any problems with this app template to contact@estimote.com
//

protocol ShowroomManagerDelegate: class {

    func showroomManager(_ showroomManager: ShowroomManager, didDetectPickupForProduct product: Product)

    func showroomManager(_ showroomManager: ShowroomManager, didDetectPutdownForProduct product: Product)

}

class ShowroomManager: NSObject, ESTTriggerManagerDelegate {

    weak var delegate: ShowroomManagerDelegate?

    private let products: [NearableID: Product]

    private let triggerManager = ESTTriggerManager()
    private var triggers = [ESTTrigger]()

    init(products: [NearableID: Product]) {
        self.products = products

        super.init()

        self.triggerManager.delegate = self

        for nearableID in products.keys {
            let motionRule = ESTMotionRule.motionStateEquals(true, forNearableIdentifier: nearableID.nearableIDString)
            let motionTrigger = ESTTrigger(rules: [motionRule], identifier: nearableID.nearableIDString)
            self.triggers.append(motionTrigger)
        }
    }

    func startUpdates() {
        for trigger in self.triggers {
            self.triggerManager.startMonitoring(for: trigger)
        }
    }

    func stoptUpdates() {
        for trigger in self.triggers {
            self.triggerManager.stopMonitoringForTrigger(withIdentifier: trigger.identifier)
        }
    }

    func triggerManager(_ manager: ESTTriggerManager, triggerChangedState trigger: ESTTrigger) {
        if let nearableID = NearableID(nearableIDString: trigger.identifier), let product = self.products[nearableID] {
            if trigger.state == true {
                self.delegate?.showroomManager(self, didDetectPickupForProduct: product)
            } else {
                self.delegate?.showroomManager(self, didDetectPutdownForProduct: product)
            }
        }
    }

}
